File: Quintic_Bezier.py
from Point import *

import numpy as np
class Quintic_Bezier:
    def __init__(self, point0, point1, point2, point3, point4, point5):
        self.point0 = point0
        self.point1 = point1
        self.point2 = point2
        self.point3 = point3
        self.point4 = point4
        self.point5 = point5

    def get_control_points(self):
        return self.point0, self.point1, self.point2, self.point3, self.point4, self.point5

    def calc_first_derivative(self, t):
        p = ((5 * ((1 - t) ** 4) * (self.point1 - self.point0)) + (20 * t * ((1 - t) ** 3) * (self.point2 - self.point1)) +
            (30 * (t ** 2) * ((1 - t) ** 2) * (self.point3 - self.point2)) + (20 * (t ** 3) * (1 - t) * (self.point4 - self.point3)) +
            (5 * (t ** 4) * (self.point5 - self.point4)))

        return p

    def calc_second_derivative(self, t):
        p = ((20 * ((1 - t) ** 3) * (self.point2 - (2 * self.point1) + self.point0)) +
             (60 * t * ((1 - t) ** 2) * (self.point3 - (2 * self.point2) + self.point1)) +
             (60 * (t ** 2) * (1 - t) * (self.point4 - (2 * self.point3) + self.point2) +
              (20 * (t ** 3) * (self.point5 - (2 * self.point4) + self.point3))))

        return p


    def calc_curvature(self, t):
        # print("d", (determinant(self.calc_first_derivative(t), self.calc_second_derivative(t))))
        if magnitude(self.calc_first_derivative(t)) == 0:
            return 0
        return (determinant(self.calc_first_derivative(t), self.calc_second_derivative(t)))/(magnitude(self.calc_first_derivative(t)) ** 3)

    def get_point(self, t):
        """Returns point in quintic bezier curve. """
        p = (((1 - t) ** 5) * self.point0 + (5 * ((1 - t) ** 4) * t * self.point1) + (10 * ((1 - t) ** 3) * (t ** 2) * self.point2) +
             (10 * ((1 - t) ** 2) * (t ** 3) * self.point3 + 5 * (1 - t) * (t ** 4) * self.point4) + ((t ** 5) * self.point5))

        return p

    def calc_arc_length(self):
        distance = 0
        prev_point = self.get_point(0);
        for t in np.arange(0, 1.01, 0.01):
            curr_point = self.get_point(t)
            distance += distance_formula(curr_point, prev_point)
            prev_point = curr_point
        return distance

File: frontend/src/Quintic_Bezier.js
// QuinticBezier.js
import { Point, distance_formula, magnitude, determinant } from "./Point.js";

export class Quintic_Bezier {
  constructor(point0, point1, point2, point3, point4, point5) {
    this.point0 = point0;
    this.point1 = point1;
    this.point2 = point2;
    this.point3 = point3;
    this.point4 = point4;
    this.point5 = point5;
  }
  
  getControlPoints() {
    return [this.point0, this.point1, this.point2, this.point3, this.point4, this.point5];
  }
  
  calcFirstDerivative(t) {
    const term1 = this.point1.subtract(this.point0).multiply(5 * Math.pow(1 - t, 4));
    const term2 = this.point2.subtract(this.point1).multiply(20 * t * Math.pow(1 - t, 3));
    const term3 = this.point3.subtract(this.point2).multiply(30 * Math.pow(t, 2) * Math.pow(1 - t, 2));
    const term4 = this.point4.subtract(this.point3).multiply(20 * Math.pow(t, 3) * (1 - t));
    const term5 = this.point5.subtract(this.point4).multiply(5 * Math.pow(t, 4));
    return term1.add(term2).add(term3).add(term4).add(term5);
  }
  
  calcSecondDerivative(t) {
    const term1 = this.point2.subtract(this.point1.multiply(2)).add(this.point0)
      .multiply(20 * Math.pow(1 - t, 3));
    const term2 = this.point3.subtract(this.point2.multiply(2)).add(this.point1)
      .multiply(60 * t * Math.pow(1 - t, 2));
    const term3 = this.point4.subtract(this.point3.multiply(2)).add(this.point2)
      .multiply(60 * Math.pow(t, 2) * (1 - t));
    const term4 = this.point5.subtract(this.point4.multiply(2)).add(this.point3)
      .multiply(20 * Math.pow(t, 3));
    return term1.add(term2).add(term3).add(term4);
  }
  
  calcCurvature(t) {
    const firstDeriv = this.calcFirstDerivative(t);
    if (magnitude(firstDeriv) === 0) {
      return 0;
    }
    const secondDeriv = this.calcSecondDerivative(t);
    return determinant(firstDeriv, secondDeriv) / Math.pow(magnitude(firstDeriv), 3);
  }
  
  getPoint(t) {
    const term0 = this.point0.multiply(Math.pow(1 - t, 5));
    const term1 = this.point1.multiply(5 * Math.pow(1 - t, 4) * t);
    const term2 = this.point2.multiply(10 * Math.pow(1 - t, 3) * Math.pow(t, 2));
    const term3 = this.point3.multiply(10 * Math.pow(1 - t, 2) * Math.pow(t, 3));
    const term4 = this.point4.multiply(5 * (1 - t) * Math.pow(t, 4));
    const term5 = this.point5.multiply(Math.pow(t, 5));
    return term0.add(term1).add(term2).add(term3).add(term4).add(term5);
  }
  
  calcArcLength() {
    let distance = 0;
    let prevPoint = this.getPoint(0);
    // Loop from t=0 to t=1 in steps of 0.01
    for (let t = 0; t <= 1.0001; t += 0.01) {
      const currPoint = this.getPoint(t);
      distance += distance_formula(currPoint, prevPoint);
      prevPoint = currPoint;
    }
    return distance;
  }
  

}
